src/schema/TlvSchema.ts

```diff
@@ -0,0 +1,242 @@
+                // No need to write the tag
+                break;
+        }
+    }
+
+    /** @override */
+    protected decodeInternal(encoded: ArrayBuffer): T {
+        return this.decodeTlv(new DataReaderLE(encoded)).value;
+    }
+
+    /** @override */
+    protected encodeInternal(value: T): ArrayBuffer {
+        const writer = new DataWriterLE();
+        this.encodeTlv(writer, value);
+        return writer.toBuffer();
+    }
+
+    /** Gets {@link ElementSize} based on the number value. */
+    protected getUnsignedIntSize(value: number | bigint) {
+        if (value < 256) {
+            return ElementSize.Byte1;
+        } else if (value < 65536) {
+            return ElementSize.Byte2;
+        } else if (value < 4294967296) {
+            return ElementSize.Byte4;
+        } else {
+            return ElementSize.Byte8;
+        }
+    }
+
+    /** Encodes an unsigned integer value. */
+    protected encodeUnsignedIntBytes(writer: DataWriterLE, value: number | bigint, size: ElementSize) {
+        if (typeof value === "number") {
+            switch (size) {
+                case ElementSize.Byte1:
+                    writer.writeUInt8(value);
+                    break;
+                case ElementSize.Byte2:
+                    writer.writeUInt16(value);
+                    break;
+                case ElementSize.Byte4:
+                    writer.writeUInt32(value);
+                    break;
+                case ElementSize.Byte8:
+                    throw new Error("64 bits number should be stored on a bigint");
+            }
+        } else {
+            switch (size) {
+                case ElementSize.Byte1:
+                    writer.writeUInt8(Number(value));
+                    break;
+                case ElementSize.Byte2:
+                    writer.writeUInt16(Number(value));
+                    break;
+                case ElementSize.Byte4:
+                    writer.writeUInt32(Number(value));
+                    break;
+                case ElementSize.Byte8:
+                    writer.writeUInt64(value);
+                    break;
+            }
+        }
+    }
+
+    /** Decodes a TLV value. */
+    protected abstract decodeTlv(reader: DataReaderLE): { value: T, tag: TlvTag};
+
+    /** Encodes a TLV value. */
+    protected abstract encodeTlv(writer: DataWriterLE, value: T, tag?: TlvTag): void;
+}
+
+/**
+ * Schema to encode an unsigned integer in TLV.
+ * 
+ * @see {@link MatterCoreSpecificationV1_0} § A.11.1
+ */
+class UnsignedNumberSchema extends TlvSchema<number | bigint> {
+    /** @override */
+    protected encodeTlv(writer: DataWriterLE, value: number | bigint, tag: TlvTag = {}): void {
+        const size = this.getUnsignedIntSize(value);
+        this.encodeControlByteAndTag(writer, ElementType.UnsignedInt | size, tag);
+        this.encodeUnsignedIntBytes(writer, value, size);
+    }
+
+    /** @override */
+    protected decodeTlv(encoded: DataReaderLE) {
```

Comment:
Could this implementation be moved to a static method in `TlvCodec`?

Reply:
Will do.